File: Lab6-C/cscd437Lab6Main.c
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>
#include <regex.h>
#include <limits.h>
#include <errno.h>
#include <openssl/sha.h>
#include "sodium.h"

struct Logs {
    char* first;
    char* second;
    char* third;
    char* fourth;
};

void logging(struct Logs*, char*);
void name(char*);
bool regexMatch(char*, char*);
void cullNewLine(char*);
void number(char*);
void openFile(char*);
void password();
void hashPassword(char*);
char* hash(char*, char*);
void reenterPassword();
void writeToOutput();
void passwordError(char*);
void writeError();
void writeToFile(char*, char*);
void numError(char*, char*);
void fileError(char*, char*);
void writeToOutput();

static struct Logs mErrors;
static bool mFirstNameCorrect, mLastNameCorrect, mNum1Correct, mNum2Correct, mInFileCorrect, mOutFileCorrect, mPasswordCorrect;
static FILE* mInput;
static FILE* mOutput;
static char* mFirst;
static char* mLast;
static char* mNum1;
static char* mNum2;
static char* mPassword;
static char* mPasswordFile;

int main(int argc, char *argv[]) {
    (void)argc;
    (void)argv;
    name("first");
    name("last");

    number("first");
    number("second");

    openFile("input");
    openFile("output");

    password();
    reenterPassword();

    writeToOutput();
    writeToFile("errors-c.txt", NULL);
    printf("Error:\n\t%s\n\t%s\n\t%s\n\t%s\n", mErrors.first, mErrors.second, mErrors.third, mErrors.fourth);
    return 0;
}

void name(char* which) {
    printf("Enter %s name: ", which);
    char* name = malloc(51);
    char* log = malloc(128);
    if(fgets(name, 51, stdin)) {
        cullNewLine(name);
        char* pattern = "^[A-Za-z,.'-]{1,50}$";
        if(regexMatch(name, pattern)) {
            if(strcmp(which, "first") == 0) {
                mFirst = name;
                mFirstNameCorrect=true;
            }
            else {
                mLast = name;
                mLastNameCorrect=true;
            }
        } else {
            asprintf(&log, "Error when reading %s name. Entered bad value '%s'.\n", which, name);
            cullNewLine(log);
            logging(&mErrors, log);
        }
    } else {
        asprintf(&log, "Error when reading %s name. Entered bad value '%s'.\n", which, name);
        cullNewLine(log);
        logging(&mErrors, log);
    }
}

void cullNewLine(char* str) {
    char* noline = NULL;
    noline = strchr(str, '\n');
    if(noline) *noline = '\0';
    else while(getchar() != '\n');
}

void logging(struct Logs* mLog, char* log) {
    if(!mLog->first) {
        mLog->first = malloc(128);
        strcpy(mLog->first, log);
    } else if(!mLog->second) {
        mLog->second = malloc(128);
        strcpy(mLog->second, log);
    } else if(!mLog->third) {
        mLog->third = malloc(128);
        strcpy(mLog->third, log);
    } else if(!mLog->fourth) {
        mLog->fourth = malloc(128);
        strcpy(mLog->fourth, log);
    } else printf("Error logs full.\n");
}

bool regexMatch(char* str, char* pattern) {
    regex_t regex;
    int regInt;
    regInt = regcomp(&regex, pattern, REG_EXTENDED);
    if(regInt != 0) return false;
    regInt = regexec(&regex, str, 0, NULL, 0);
    regfree(&regex);
    if(regInt == 0) return true;
    return false;
}

void number(char* which) {
    printf("Enter %s integer: ", which);
    char* number = malloc(48);
    if(fgets(number, 48, stdin)) {
        cullNewLine(number);
        char* pattern = "^(-?[0-9]+){1,11}$";
        if(regexMatch(number, pattern)) {
            char* endptr;
            long num = strtol(number, &endptr, 10);
            if(endptr != number) {
                if(errno == 0) {
                    if(num >= INT_MIN && num <= INT_MAX) {
                        if(strcmp(which, "first") == 0) {
                            mNum1 = number;
                            mNum1Correct=true;
                        }
                        else {
                            mNum2 = number;
                            mNum2Correct;
                        }
                    } else numError(which, number);
                } else numError(which, number);
            } else numError(which, number);
        } else numError(which, number);
    } else numError(which, number);
}

void numError(char* which, char* number) {
    char* log = malloc(128);
    asprintf(&log, "Error when reading %s number. Entered bad value '%s'.\n", which, number);
    cullNewLine(log);
    logging(&mErrors, log);
}

void openFile(char* mode) {
printf("Enter %s file name: ", mode);
    char* fname = malloc(256);
    char* rorw;
    if(fgets(fname, 256, stdin)) {
        cullNewLine(fname);
        char* pattern = "^[0-9A-Za-z,.'-]+{1,200}\\.([A-Za-z]+){1,20}$";
        if(regexMatch(fname, pattern)) {
            if(strcmp(mode, "input")==0) rorw = "r";
            else rorw = "w";
            FILE* file = fopen(fname, rorw);
            if(file) {
                if(strcmp(rorw, "r")==0) {
                    mInput = file;
                    mInFileCorrect = true;
                } else {
                    mOutput = file;
                    mOutFileCorrect = true;
                }
            } else fileError(mode, fname);
        } else fileError(mode, fname);
    } else fileError(mode, fname);
}


void fileError(char* mode, char* fname) {
    char* log = malloc(356);
    asprint(&log, "Error when reading %s file. Entered bad value '%s'.\n", mode, fname);
    cullNewLine(log);
    logging(&mErrors, log);
}

void password() {
	printf("Enter password: ");
	char* password=(char*)malloc(129*sizeof(char));
    char* log = malloc(128);
	if(fgets(password, 129*sizeof(char), stdin)) {
		cullNewLine(password);
        char* pattern = "^((?!\\0).){1,128}$";
        if(regexMatch(password, pattern)) {
            mPasswordCorrect=true;
            hashPassword(password);
        }
        else {
            asprintf(&log, "Error when reading password. Entered bad value '%s'.\n", password);
            cullNewLine(log);
            logging(&mErrors, log);
        }
	}
    else {
        asprintf(&log, "Error when reading password. Entered bad value '%s'.\n", password);
        cullNewLine(log);
        logging(&mErrors, log);
    }
}

void hashPassword(char* password) {
    char salt[32];
    randombytes_buf(salt, 32);
    char* h=salt;
    h=hash(h, password);

    writeToFile("password.txt", hash);
}

char* hash(char* hash, char*password) {
    char data[] = password;
    char h[SHA512_DIGEST_LENGTH];
    SHA512(data, sizeof(data) - 1, h);
    hash=hash+(char)h;

    return hash;
}

void reenterPassword() {
    if(mPasswordCorrect) {
        printf("Reenter password: ");
	    char* password=(char*)malloc(129*sizeof(char));
        char* log = malloc(128);
	    if(fgets(password, 129*sizeof(char), stdin)) {
            cullNewLine(password);
            FILE* file=fopen(mPasswordFile, "w+");
            char firstPass[64];
            fscanf(file, "%[^\n]", firstPass);
            char* salt=strtok(firstPass, "\\]")+(char)"\\]";
            char* h=hash(salt, password);
            bool same=true;
            int cur=0;
            int i=strnlen(h, 65);
            while(cur<i) {
                same=firstPass[i]==h[i];
                cur++;
                if(!same) break;
            }
            if(same) printf("Passwords are equivalent.");
            else passwordError(password);
        }
        else passwordError(password);
    } else printf("Previous password not formatted correctly, can't re-enter password.\n");
}

void writeToFile(char* fileName, char* hash) {
    FILE* file = NULL;
    file=fopen(fileName, "w+");
    char* fname=fileName;
    while(access( file, F_OK ) != -1 ) {
        fname=0+fname;
        file=fopen(fname, "w+");
    }
    if(fileName=="password.txt") mPasswordFile=fname;

    if(fileName=="errors-c.txt") writeError();
    else fputs(hash, file);
    close(file);
}

static void writeError() {
    FILE* fout;
    fout = fopen("errors-c.txt", "w+");
    fprintf(fout, "%s", mErrors.first);
    fprintf(fout, "%s", mErrors.second);
    fprintf(fout, "%s", mErrors.third);
    fprintf(fout, "%s", mErrors.fourth);
    fclose(fout);
}


void passwordError(char* password) {
    char* log = malloc(128);
    asprintf(&log, "The password '%s' does not match the first password entered.\n", password);
    cullNewLine(log);
    logging(&mErrors, log);
}

void writeToOutput() {
    char* log = malloc(128);
    if(mOutFileCorrect) {
        FILE* file=fopen(mOutput, "w+");
        if(mFirstNameCorrect) {
            fputs(mFirst+' ', file);
            if(!mLastNameCorrect) fputs('\n', file);
        }
        if(mLastNameCorrect) fputs(mLast+'\n', file);
        if(!(mFirstNameCorrect&&mLastNameCorrect)) {
            asprintf(&log, "Could not write name to output file; valid names not given.\n");
            cullNewLine(log);
            logging(&mErrors, log);
        }

        if(mNum1Correct&&mNum2Correct) {
            char* endptr;
            long num1 = strtol(mNum1, &endptr, 10);
            long num2 = strtol(mNum2, &endptr, 10);
            long add=num1+num2;
            long mult=num1*num2;
            fputs(num1+'+'+num2+'='+add+'\n', file);
            fputs(num1+'*'+num2+'='+mult+'\n', file);
        } else {
            asprintf(&log, "Could not write integer addition/multiplication to output file; one or both integers not valid.\n");
            cullNewLine(log);
            logging(&mErrors, log);
            fputs('\n', file);
        }

        if(mInFileCorrect) {
            char* s="Input file contents:\n";
            fputs(s, file);
            FILE* input=fopen(mInput, "w+");

            char* line = NULL;
            size_t len = 0;
            ssize_t read;

            while ((read = getline(&line, &len, input)) != -1) fputs(line, file);
        } else {
            asprintf(&log, "Could not write input file contents to output file; valid filename not given.\n");
            cullNewLine(log);
            logging(&mErrors, log);
        }
    } else {
        asprintf(&log, "Could not write to output file; valid filename not given.\n");
        cullNewLine(log);
        logging(&mErrors, log);
    }


}

File: Lab6-Java/CSCD437Lab6Main.java
/*
 * Ron Robinson III - Katherine Bozin
 * Tech Support
 * CSCD437-U20
*/

import java.io.*;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CSCD437Lab6Main {

	private static ArrayList<String> mErrors;
	private static boolean mFirstNameCorrect, mLastNameCorrect, mNum1Correct, mNum2Correct, mInFileCorrect, mOutFileCorrect, mPasswordCorrect;
	private static File mInput, mOutput;
	private static String mFirst, mLast, mNum1, mNum2, mPasswordFile;
	
	public static void main(String[] args) {
		mErrors=new ArrayList<>();
		mFirstNameCorrect=false;
		mLastNameCorrect=false;
		mNum1Correct=false;
		mNum2Correct=false;
		mInFileCorrect=false;
		mOutFileCorrect=false;
		mPasswordCorrect=false;
		mInput=null;
		mOutput=null;

		Scanner in = new Scanner(System.in);
		name("first", in);
		name("last", in);

		number("first", in);
		number("second", in);
		
		openFile("input", in);
		openFile("output", in);

		password(in);
		reenterPassword(in);

		writeToOutput();
		writeToFile("errors-java.txt", null, null);
		
		in.close();
	}

	private static void name(final String which, final Scanner in) {
		System.out.print("Enter "+which+" name: ");
		String name=in.nextLine();

		String namePattern = "^[A-Za-z,.'-]{1,50}$";
		Pattern pattern = Pattern.compile(namePattern);
		Matcher match = pattern.matcher(name);
		if(match.matches()) {
			if(which.equals("first")) {
				mFirst=name;
				mFirstNameCorrect=true;
			}
			else {
				mLast=name;
				mLastNameCorrect=true;
			}
		} else {
			mErrors.add("Error when reading "+which+" name. Entered bad value '"+name+"'.");
		}
	}
	
	private static void number(final String which, final Scanner in) {
		System.out.print("Enter "+which+" integer: ");
		String number=in.nextLine();

		String numberPattern = "^(-?\\d+){1,11}$";
		Pattern pattern = Pattern.compile(numberPattern);
		Matcher match = pattern.matcher(number);

		if(match.matches()) {
			BigInteger bigInt=new BigInteger(number);
			if(bigInt.compareTo(BigInteger.valueOf(Integer.MIN_VALUE)) >= 0 &&
					bigInt.compareTo(BigInteger.valueOf(Integer.MAX_VALUE)) <= 0) {
				if(which.equals("first")) {
					mNum1=number;
					mNum1Correct=true;
				}
				else {
					mNum2=number;
					mNum2Correct=true;
				}
			} else {
				mErrors.add("Error when reading "+which+" number. Entered bad value '"+number+"'.");
			}
		} else {
			mErrors.add("Error when reading "+which+" number. Entered bad value '"+number+"'.");
		}

	}

	private static void openFile(final String mode, final Scanner in) {
		System.out.print("Enter "+mode+" file name: ");
		String fileName = in.nextLine();

		String filePattern = "^[\\w,-. ]{1,200}\\.([A-Za-z]){1,20}$";
		Pattern pattern = Pattern.compile(filePattern);
		Matcher match = pattern.matcher(fileName);

		if (match.matches()) {
			File file = new File(fileName);
			if (file.exists()) {
				if (mode.equals("input") && file.canRead()) {
					mInput=file;
					mInFileCorrect=true;
				}
				else if (mode.equals("output") && file.canWrite()) {
					mOutput=file;
					mOutFileCorrect=true;
				} else fileError(mode, fileName);
			} else fileError(mode, fileName);
		} else fileError(mode, fileName);
	}

	private static void fileError(String mode, String fileName) {
		mErrors.add("Error when reading "+mode+" file. Entered bad value '"+fileName+"'.");
	}
	
	private static void password(final Scanner in) {
		System.out.print("Enter password: ");
		String password = in.nextLine();

		String passwordPattern = "^((?!\\\\0).){1,128}$";
		Pattern pattern = Pattern.compile(passwordPattern);
		Matcher match = pattern.matcher(password);

		if(match.matches()) {
			mPasswordCorrect=true;
			hashPassword(password);
		}
		else {
			mErrors.add("Error when reading password. Entered bad value '"+password+"'.");
		}
	}

	private static void hashPassword(String password) {
		String hash;
		SecureRandom random = new SecureRandom();
		byte[] salt = new byte[16];
		random.nextBytes(salt);
		hash=Arrays.toString(salt);
		hash=hash(hash, password);

		writeToFile("password.txt", hash, salt);
	}

	private static String hash(String hash, String password) {
		try {
			MessageDigest md = MessageDigest.getInstance("SHA-512");
			byte[] messageDigest = md.digest(password.getBytes());
			BigInteger bigInt = new BigInteger(1, messageDigest);
			hash = hash+bigInt.toString(16);
			while (hash.length()<32) hash = "0".concat(hash);
		} catch (NoSuchAlgorithmException e) { mErrors.add("NoSuchAlgorithmException in method 'hash'. "+e.getMessage()); }

		return hash;
	}

	private static void writeToFile(String fileName, String hash, byte[] salt) {
		File file=new File(fileName);
		String fname=fileName;
		while(file.exists()) {
			fname="0".concat(fname);
			file=new File(fname);
		}
		if(fileName.equals("password.txt")) mPasswordFile=fname;
		try { file.createNewFile(); }
		catch (IOException e) { mErrors.add("IOException in method 'writeToFile'. "+e.getMessage()); }

		BufferedWriter fout = null;
		try { fout = new BufferedWriter(new FileWriter(file)); }
		catch (IOException e) { mErrors.add("IOException in method 'writeToFile'. "+e.getMessage()); }

		assert fout != null;
		if(fileName.equals("errors-java.txt")) writeError(fout);
		else writePassword(fout, hash);
		try { fout.close(); }
		catch (IOException e) { mErrors.add("IOException in method 'writeToFile'. "+e.getMessage()); }
	}

	private static void writePassword(BufferedWriter fout, String hash) {
		try { fout.write(hash); }
		catch (IOException e) { mErrors.add("IOException in method 'writePassword'. "+e.getMessage()); }
	}

	private static void writeError(BufferedWriter fout) {
		try {
			for (String mError : mErrors) fout.write(mError + "\n");
			if (mErrors.isEmpty()) fout.write("No errors.");
		} catch (IOException e) { mErrors.add("IOException in method 'writeError'. "+e.getMessage()); }
	}

	private static void reenterPassword(Scanner in) {
		if(mPasswordCorrect) {
			System.out.print("Re-enter password: ");
			String password=in.nextLine();
			String hash="", firstPass="", salt="";

			FileInputStream fis = null;
			try { fis = new FileInputStream(mPasswordFile); }
			catch (FileNotFoundException e) { mErrors.add("FileNotFoundException in method 'reenterPassword'. "+e.getMessage()); }

			assert fis!=null;
			BufferedReader br = new BufferedReader(new InputStreamReader(fis));
			try {
				firstPass=br.readLine();
				salt=firstPass.split("\\]")[0]+"]";
				hash=hash(salt, password);
			} catch (IOException e) { mErrors.add("IOException in method 'reenterPassword'. "+e.getMessage()); }

			try { br.close(); }
			catch (IOException e) { mErrors.add("IOException in method 'reenterPassword'. "+e.getMessage()); }

			boolean same=true;
			int i=hash.length(), cur=0;
			if(i!=firstPass.length()) passwordError(password);
			else {
				while(cur<i) {
					same=firstPass.charAt(cur)==hash.charAt(cur);
					cur++;
					if(!same) break;
				}

				if(same) System.out.println("Passwords are equivalent.");
				else passwordError(password);
			}
		}
		else System.out.println("Previous password not formatted correctly, can't re-enter password.");

	}

	private static void passwordError(String password) {
		System.out.println("The password '"+password+"' does not match the first password entered.");
		mErrors.add("Error when rereading password. Entered bad value '"+password+"'. Password does not match original password.");
	}

	private static void writeToOutput() {
		if(mOutFileCorrect) {
			BufferedWriter fout = null;
			try { fout = new BufferedWriter(new FileWriter(mOutput)); }
			catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			assert fout!=null;

			if(mFirstNameCorrect) {
				try {
					fout.write(mFirst+" ");
					if(!mLastNameCorrect) fout.write("\n");
				}
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			}
			if(mLastNameCorrect) {
				try { fout.write(mLast+"\n"); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			}
			if(!(mFirstNameCorrect&&mLastNameCorrect)) {
				mErrors.add("Could not write name to output file; valid names not given.");
			}

			if(mNum1Correct&&mNum2Correct) {
				BigInteger bigIntOne=new BigInteger(mNum1);
				BigInteger bigIntTwo=new BigInteger(mNum2);
				try { fout.write(mNum1+"+"+mNum2+"="+bigIntOne.add(bigIntTwo)+"\n"+mNum1+"*"+mNum2+"="+bigIntOne.multiply(bigIntTwo)+"\n"); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			} else {
				mErrors.add("Could not write integer addition/multiplication to output file; one or both integers not valid.");
				try { fout.write("\n"); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			}

			if(mInFileCorrect) {
				try { fout.write("Input file contents:\n"); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
				FileInputStream fis = null;
				try { fis = new FileInputStream(mInput); }
				catch (FileNotFoundException e) { mErrors.add("FileNotFoundException in method 'writeToOutput'. "+e.getMessage()); }

				assert fis!=null;
				BufferedReader in = new BufferedReader(new InputStreamReader(fis));
				String line = null;
				while (true) {
					try { if ((line = in.readLine()) == null) break; }
					catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
					try { fout.write(line+"\n"); }
					catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
				}
				try { fis.close(); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
				try { in.close(); }
				catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
			} else {
				mErrors.add("Could not write input file contents to output file; valid filename not given.");
			}
			try { fout.close(); }
			catch (IOException e) { mErrors.add("IOException in method 'writeToOutput'. "+e.getMessage()); }
		} else {
			mErrors.add("Could not write to output file; valid filename not given.");
		}
	}
}